### PR TITLE
Add check against NULL before dereferencing pointer

### PR DESCRIPTION
**Pull request by valeriyvan**
Condition following fix suggests pointer is expected sometimes to be NULL.

### CODE REVIEW

File: lib/IDE/Formatting.cpp [hunk 1]
```diff
@@ -2485,7 +2485,7 @@ class FormatWalker : public ASTWalker {
   getIndentContextFrom(CaptureListExpr *CL,
                        SourceLoc ContextLoc = SourceLoc()) {
     AbstractClosureExpr *CE = CL->getClosureBody();
-    BraceStmt *BS = CE->getBody();
+    BraceStmt *BS = CE ? CE->getBody() : NULL;
```
Comment (AnthonyLatsis) [inline lib/IDE/Formatting.cpp:2488]: None of these can ever be a null pointer, but the invariants are neglected. Can we assert this for the closure upon `CaptureListExpr` construction, and remove the always true `hasBody()` check from `AbstractClosureExpr::getBody`? This also seems to be the only remaining use of `hasBody()`, so we may as well just delete it.

Comment (valeriyvan) [reply lib/IDE/Formatting.cpp:2488]: `lib/SIL/IR/SILProfiler.cpp` still uses `hasBody()` in assert. But if `hasBody()` always returns `true` this assert could also be removed.

Comment (valeriyvan) [reply lib/IDE/Formatting.cpp]: done

File: lib/IDE/Formatting.cpp [hunk 2]
```diff
@@ -2485,8 +2485,9 @@ class FormatWalker : public ASTWalker {
   getIndentContextFrom(CaptureListExpr *CL,
                        SourceLoc ContextLoc = SourceLoc()) {
     AbstractClosureExpr *CE = CL->getClosureBody();
+    assert(CE);
```
Comment (AnthonyLatsis) [inline lib/IDE/Formatting.cpp:2488]: Not here, in the `CaptureListExpr` constructor; sorry if I wasn’t clear.

File: include/swift/AST/Expr.h [hunk 3]
```diff
@@ -3863,9 +3863,6 @@ class AbstractClosureExpr : public DeclContext, public Expr {
   /// Only valid when \c hasSingleExpressionBody() is true.
   Expr *getSingleExpressionBody() const;
 
-  /// Whether this closure has a body
-  bool hasBody() const;
-
   /// Returns the body of closures that have a body
   /// returns nullptr if the closure doesn't have a body
   BraceStmt *getBody() const;
```
Comment (AnthonyLatsis) [inline include/swift/AST/Expr.h:3868]: The mention of not having a body can go away now.

File: lib/IDE/Formatting.cpp [hunk 4]
```diff
@@ -2485,8 +2485,9 @@ class FormatWalker : public ASTWalker {
...
     BraceStmt *BS = CE->getBody();
-    if (!CE || !BS)
+    if (!BS)
```
Comment (AnthonyLatsis) [inline lib/IDE/Formatting.cpp:2490]: This check can be removed now as well.

File: lib/IDE/Formatting.cpp (hunk 1 above)
Comment (AnthonyLatsis) [reply lib/IDE/Formatting.cpp]: > are you sure I should remove `hasBody()` completely?

At least I’d vote for it. @etcwilde What was the motivation for adding it in c57b80a3c88cb2eaf5825c0c22848f0d4ebab70c?

**Comment by etcwilde** [comment]
>> are you sure I should remove hasBody() completely?

> At least I’d vote for it. @etcwilde What was the motivation for adding it in https://github.com/apple/swift/commit/c57b80a3c88cb2eaf5825c0c22848f0d4ebab70c?

I don't remember exactly why I added it. I think it was to mirror another API? Maybe the `FuncDecl` APIs?
Anyway, yeah, it doesn't look terribly useful.

IIRC, the calls underlying `getBody` at least used to be able to return `nullptr` if you were querying it during type checking since they bodies hadn't necessarily been set yet. If something failed to type check, you will probably still want to be defensive here.

**Comment by xedin** [comment]
@AnthonyLatsis 's suggestions make sense to me.

**Comment by AnthonyLatsis** [comment]
> IIRC, the calls underlying `getBody` at least used to be able to return `nullptr` if you were querying it during type checking since they bodies hadn't necessarily been set yet. If something failed to type check, you will probably still want to be defensive here.

Good point.

File: lib/IDE/Formatting.cpp (hunk 2 above)
Comment (valeriyvan) [reply lib/IDE/Formatting.cpp:2488]: Sorry, I get lost. By constructor do you mean `CaptureListExpr::create`? If so what should I assert there? `assert(mem);`?

Comment (AnthonyLatsis) [reply lib/IDE/Formatting.cpp:2488]: > By constructor do you mean `CaptureListExpr::create`?

Yes, but preferably the actual constructor, `CaptureListExpr::CaptureListExpr`. It’s defined in the header.

> If so what should I assert there?

`assert(closureBody)`.

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Review by xedin — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
RxSwift failure is unrelated, merging.

**Pull request merged**